### PR TITLE
Add support for an `ilab` global config

### PR DESCRIPTION
**Pull request by cdoern**
When the env var `ILAB_GLOBAL_CONFIG` is set, that config is used as the
template to render the ~/.config/instructlab/config.yaml user space
config. The interactive mode is skipped when calling `ilab config init`.
This works at the `ilab` main entrypoint, this can be used to specify a
different configuration file too when running any commands.

This ensures defaults like taxonomy repos, base models, etc, are honored
from the global config.

Co-authored-by: Sébastien Han <seb@redhat.com>
Signed-off-by: Charlie Doern <cdoern@redhat.com>

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by leseb** [comment]
I don't think this can work until https://github.com/instructlab/instructlab/issues/1800 is resolved.

File: src/instructlab/config/init.py
Comment (leseb) [reply src/instructlab/config/init.py:83]: It's working for me, I'm going to push an update to this branch now.

Comment (jaideepr97) [reply src/instructlab/config/init.py]: are these the only fields we care about copying over? It's still not working for me but that might be because I'm editing fields like `num_cpu` since my undersanding is we want to take the entire config as is

Comment (jaideepr97) [reply src/instructlab/config/init.py]: nvm, I suppose the entire golden config should be getting stored in the `ctx` obj so I'm not sure why my changes aren't getting reflected

Comment (leseb) [reply src/instructlab/config/init.py]: `ILAB_GLOBAL_CONFIG=config-leseb.yaml ilab config init` works for me, for now I just edited`general.log_level` and it's reflected in the generated config file.

I'm writing test!

File: src/instructlab/config/init.py
Comment (leseb) [reply src/instructlab/config/init.py:80]: correct!

Comment (cdoern) [reply src/instructlab/config/init.py:80]: also I think after speaking w/ @mairin we want running something like `ilab model chat` to just _work_ if you have the `ILAB_GLOBAL_CONFIG` env var set for those types of users. So technically we will tell them to run `ilab config init` but if they skip it, it will pull from the golden cfg? 

This may be confusing though

**Review by RobotSail — APPROVED** [review]
LGTM

**Review by jaideepr97 — APPROVED** [review]

**Comment by booxter** [comment]
I don't know if there was a design decision already made  somewhere to implement it in this particular way. (If so, it would help to see a link to where it was discussed.) So the comment may or may not be helpful. Thanks for consideration anyway.

---

A common "golden config" approach used by traditional Unix tools and elsewhere is:

1. put a "golden" config somewhere in `/usr/share/`, e.g. installed with a package;
2. let the system admin to adjust it in `/etc/` by **overriding** options from the golden config defaults; (meaning: listing changes only)
3. let user to adjust it in `$HOME/.local/config/` by **overriding** options from the both the golden and system-admin configs. (also listing changes only).
4. finally, apply the in-app (embedded) defaults if no prior config overrode them.

This enables a hierarchy of decisions (from packager to system admin to user). This also allows to rely on defaults defined in the golden config as well as in the app.

---

AFAIU here the approach is to have the "golden" config **copied** into `$HOME`, meaning that any changes to system-level / package defaults won't be reflected in the user config. This may an issue in the future when a packager would like to propagate a better setting to users automatically.

I wonder if we could **merge** several configs in runtime instead of copying it once on `init`.

---

Let me know what you think.

**Comment by bjhargrave** [comment]
> 3\. let user to adjust it in `$HOME/.local/config/`

Do you mean `$XDG_CONFIG_HOME` (defaults to `$HOME/.config/`) ?

**Comment by booxter** [comment]
@bjhargrave Yes I meant the XDG thingy. 👍

**Comment by jaideepr97** [comment]
This change could probably also use a changelog update

File: src/instructlab/config/init.py
Comment (booxter) [inline src/instructlab/config/init.py:92]: (No action required) This function becomes a bit too long to my taste. I'd suggest to split out the paths for golden config env (lines 88-91) and "default" (lines 93-147) into separate functions, and call them here. You'd have:

```
if param_source == ...:
    model_path, taxonomy_path, taxonomy_base, cfg = get_params_from_env(cfg)
else:
    model_path, taxonomy_path, taxonomy_base, cfg = get_params_default(cfg, ...)

...

if cfg is None:
    ...
```

It would be somewhat easier to consume the code then.

File: src/instructlab/config/init.py
Comment (booxter) [inline src/instructlab/config/init.py:91]: I [think](https://github.com/instructlab/instructlab/pull/1847#discussion_r1690148608) this is never a `None`, so you could probably move `cfg = get_default_config()` from line 150 under the `else` branch for "default" (non-golden-env) case.

File: src/instructlab/configuration.py
Comment (booxter) [inline src/instructlab/configuration.py:617]: when is `config_file is None`? This looks like it comes from the click option that has the default value set to a string:

```
@click.option(
    "--config",
    "config_file",
    type=click.Path(),
    default=cfg.DEFAULTS.CONFIG_FILE,
    envvar=cfg.DEFAULTS.ILAB_GLOBAL_CONFIG,
    show_default=True,
    help="Path to a configuration file.",
)
```

I reverted the change to see which test case fails, but none does. Is this needed?

File: tests/test_lab_config.py
Comment (booxter) [inline tests/test_lab_config.py:30]: nit: There are `logging` module symbols for all levels that you could use here instead of strings.

File: tests/test_lab_config.py
Comment (booxter) [inline tests/test_lab_config.py:39]: I think our CliRunner does not capture stderr (mix_stderr=True by default). (I think this should be changed, but that's beyond the scope for this PR.)

But I think you actually care to see the stdout if it fails?

File: tests/test_lab_config.py
Comment (booxter) [inline tests/test_lab_config.py:41]: I probably miss something, but why do you even need to extract the path from the output? Can't you just read the file from the CONFIG_DIR?

```
extracted_path = pathlib.Path(configuration.DEFAULTS.CONFIG_FILE)
```

This whole block seems unnecessary.

File: tests/test_lab_config.py
Comment (booxter) [inline tests/test_lab_config.py:31]: (No action required) This test assumes that the default is NOT "debug". Could we pick some other setting in the config that surely won't ever have our custom value as the default? I dunno, set `chat.max_tokens` to `123456`. (This will also allow us to NOT import `logging` to access the `DEBUG` symbol.)

File: tests/test_lab_config.py
Comment (booxter) [inline tests/test_lab_config.py:60]: nitty-nit: unindent since you no longer need to hold the file descriptor after line 59.

File: tests/test_lab_config.py
Comment (booxter) [inline tests/test_lab_config.py:65]: I assume that `tmp_path` fixture will remove the directory?

File: tests/test_lab_config.py
Comment (booxter) [inline tests/test_lab_config.py:82]: so what's the difference between the two cases? This line only? Can we make these into two variants of the same base test case? This is very redundant.

**Review by booxter — CHANGES_REQUESTED** [review]
The test cases need some love. Otherwise, the PR does what it promises to do and can probably go forward. Though I think it's not the best approach to manage config files overall (see the other comment). (That said, an alternative can probably be adopted later with relatively minor pain regardless whether this PR merges or not.)

To reiterate, once test cases are cleaned up, I am ok with merging it.

File: tests/test_lab_config.py
Comment (booxter) [reply tests/test_lab_config.py:31]: Pff. Nevermind, I just noticed the `assert INFO` above. Then at least the test case won't silently become no-op if we ever change the default. But the other argument stands I think.

File: src/instructlab/config/init.py
Comment (leseb) [reply src/instructlab/config/init.py:79]: Sorry this belongs below actually. Thanks.

File: src/instructlab/configuration.py
Comment (leseb) [reply src/instructlab/configuration.py:617]: It was None because at one point the default got removed. Good catch, thanks.

File: tests/test_lab_config.py
Comment (leseb) [reply tests/test_lab_config.py:65]: I forgot that the file gets deleted after the third run or something.

**Comment by booxter** [comment]
@leseb I will report a RFE issue later to revisit configuration procedure (if it's not already reported somewhere - will check). This PR can stand on its own as a short term solution.

**Comment by leseb** [comment]
> @leseb I will report a RFE issue later to revisit configuration procedure (if it's not already reported somewhere - will check). This PR can stand on its own as a short term solution.

Thanks, I'll remove the 'hold' label and let the bot merge it then.

File: src/instructlab/config/init.py
Comment (booxter) [inline src/instructlab/config/init.py:126]: methinks it's always non-None and it always has config?

File: src/instructlab/config/init.py
Comment (booxter) [inline src/instructlab/config/init.py:155]: You could return `cfg` since you already have it. That said, I think you could raise an exception and catch it outside as an indication of exit state, instead of reusing return value for indication about this except-ional situation.

File: tests/test_lab_config.py
Comment (booxter) [inline tests/test_lab_config.py:53]: pytest parametrize could probably be used here.

**Review by booxter — COMMENTED** [review]
Minor things that shouldn't block the PR.

File: src/instructlab/config/init.py
Comment (booxter) [reply src/instructlab/config/init.py:126]: Is it because the `typing.Optional` used for the parameter?

File: src/instructlab/config/init.py
Comment (leseb) [reply src/instructlab/config/init.py]: Addressed, both.

File: tests/test_lab_config.py
Comment (leseb) [reply tests/test_lab_config.py:63]: Done.

File: src/instructlab/config/init.py
Comment (leseb) [reply src/instructlab/config/init.py:126]: Yes. I think the error was:

```
src/instructlab/config/init.py:125: error: Item "None" of "Any | None" has no attribute "config"
```

File: tests/test_lab_config.py
Comment (booxter) [inline tests/test_lab_config.py:66]: nit: at this point, having a separate `setup_and_invoke_test` function is redundant and you could merge the body here.

**Review by booxter — APPROVED** [review]

File: tests/test_lab_config.py
Comment (leseb) [reply tests/test_lab_config.py:66]: argh, right!

**Review by booxter — APPROVED** [review]
Piece of cake. I will check the mypy thingy after this merges. Kudos!

**Review by booxter — APPROVED** [review]

**Pull request merged**
